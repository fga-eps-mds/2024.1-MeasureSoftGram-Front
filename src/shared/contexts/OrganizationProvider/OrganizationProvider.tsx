import React, { createContext, useState, useContext, ReactNode, useMemo, useEffect } from 'react';
import { Organization } from '@customTypes/organization';
import { organizationQuery } from '@services/organization';
import { toast } from 'react-toastify';
import { useAuth } from '@contexts/Auth';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization | null;
  currentOrganizations: Organization[];
  setCurrentOrganizations: (organizations: Organization[]) => void;
  organizationList: Organization[];
  isLoading: boolean;
  fetchOrganizations: (forceFetch?: boolean) => void;
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const { session } = useAuth();
  const [currentOrganizations, setCurrentOrganizations] = useState<Organization[]>([]);
  const [currentOrganization, setCurrentOrganization] = useState<Organization | null>(null);
  const [organizationList, setOrganizationList] = useState<Organization[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchOrganizations = async (forceFetch?: boolean) => {
    if (!session && !forceFetch) return;
    setIsLoading(true);
    try {
      const result = await organizationQuery.getAllOrganization();
      if (result.type === 'success') {
        const organizations = result.value.map(org => ({
          id: org.id ?? '',
          name: org.name,
          description: org.description ?? '',
          url: org.url ?? '',
          products: org.products ?? [],
          key: org.key ?? ''
        }));
        setOrganizationList(organizations);
      } else {
        toast.error("Erro ao carregar organizações.");
      }
    } catch (error) {
      console.error("Failed to fetch organizations:", error);
      toast.error("Erro ao carregar organizações. Por favor, tente novamente.");
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchOrganizations();
  }, []);

  useEffect(() => {
    if (organizationList.length > 0 && currentOrganizations.length === 0) {
      setCurrentOrganizations([organizationList[0]]);
    }
  }, [organizationList, currentOrganizations]);

  useEffect(() => {
    if (currentOrganizations.length > 0) {
      setCurrentOrganization(currentOrganizations[0]);
    } else {
      setCurrentOrganization(null);
    }
  }, [currentOrganizations]);

  const value = useMemo(() => {
    return {
      currentOrganization,
      currentOrganizations,
      setCurrentOrganizations,
      organizationList,
      isLoading,
      fetchOrganizations
    };
  }, [currentOrganization, currentOrganizations, organizationList, isLoading]);

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationProvider');
  }

  return context;
}
