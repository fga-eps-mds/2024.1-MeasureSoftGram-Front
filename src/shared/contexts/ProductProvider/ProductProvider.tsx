import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

export interface IProductContext {
  currentProduct?: Product | null;
  setCurrentProduct: (product: Product | null) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

export const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | null | undefined>(undefined); // Initialize with undefined
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      if (!currentOrganization) {
        updateProductList([]);
        return;
      }

      const result = await productQuery.getAllProducts(currentOrganization.id);
      if (result.status === 200) {
        updateProductList(result.data.results);
      }
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      setCurrentProduct(null);
      loadAllProducts();
    }
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList,
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (!context) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}
