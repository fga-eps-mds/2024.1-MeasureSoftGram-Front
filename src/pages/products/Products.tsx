import React, { useEffect, useState } from 'react';
import Head from 'next/head';
import {
  Box,
  Button,
  Container,
  Grid,
  Typography,
  IconButton,
  Link,
  TextField,
} from '@mui/material';
import AddIcon from '@mui/icons-material/Add';
import { NextPageWithLayout } from '@pages/_app.next';
import { useProductContext } from '@contexts/ProductProvider';
import { useOrganizationContext } from '@contexts/OrganizationProvider';
import getLayout from '@components/Layout';
import CardNavigation from '@components/CardNavigation';
import { Product } from '@customTypes/product';
import useRequireAuth from '@hooks/useRequireAuth';
import { Organization } from '@customTypes/organization';
import WarningIcon from '@mui/icons-material/Warning';
import { useTranslation } from 'react-i18next';
import Skeleton from './components/Skeleton';
import ScrollableList from './components/ScrollableList/index';

const Products: NextPageWithLayout = () => {
  const { t: tp } = useTranslation('product');
  const { t: to } = useTranslation('organization');

  useRequireAuth();

  const {
    organizationList,
    currentOrganization,
    setCurrentOrganizations,
    isLoading: isLoadingOrganizations
  } = useOrganizationContext();
  const { productsList } = useProductContext();
  const [filteredProducts, setFilteredProducts] = useState<Product[]>(
    productsList ?? []
  );

  useEffect(() => {
  }, [currentOrganization]);

  useEffect(() => {
    if (productsList !== undefined) setFilteredProducts(productsList);
  }, [productsList]);

  const handleSelectedOrganization = (organization: Organization) => {
    const organizationId = organization.id
    if (currentOrganization?.id === organizationId) {
      setCurrentOrganizations([]);
    } else if (organizationList?.length) {
      const selectedOrganization = organizationList.find(
        (org) => org.id === organizationId
      );
      if (selectedOrganization) {
        setCurrentOrganizations([selectedOrganization]);
      }
    }
  };

  const handleProductFilter = (name: string) => {
    if (name == null || name === '') {
      setFilteredProducts(productsList!);
      return;
    }
    const filtered =
      productsList?.filter((product) =>
        product.name.toLowerCase().includes(name.toLowerCase())
      ) ?? [];
    setFilteredProducts(filtered);
  };

  if (!productsList) {
    return (
      <Container>
        <Skeleton />
      </Container>
    );
  }
  const boxSizing = 'border-box'
  return (
    <>
      <Head>
        <title> {tp('title')} </title>
      </Head>

      <Container maxWidth="lg" style={{ width: '100%' }}>
        <Box display="flex" flexDirection="column"
          height="60%"
        >

          <Box
            data-testid="organization-box"
            sx={{
              marginTop: '40px',
              padding: '20px 36px',
              overflowX: 'auto',
              overflowY: 'hidden',
              position: 'relative',
            }}
            style={{
              maxHeight: '120px',
            }}
          >
            <Box
              display="flex"
              flexDirection="row"
              style={{
                gap: '1rem',
                padding: '0',
                marginLeft: '-36px',
                boxSizing,
              }}
              sx={{
                '&::-webkit-scrollbar': {
                  height: '8px',
                },
                '&::-webkit-scrollbar-track': {
                  background: '#f0f0f0',
                },
                '&::-webkit-scrollbar-thumb': {
                  background: '#b0b0b0',
                  borderRadius: '4px',
                },
                scrollbarColor: '#b0b0b0 #f0f0f0',
                scrollbarWidth: 'thin',
              }}
            />
          </Box>

          {isLoadingOrganizations ? <Box
            style={{
              border: '1px solid #2B4D6F',
              borderRadius: '10px',
              boxSizing,
            }}> <Skeleton /> </Box> :

            <Box
              display="flex"
              flexDirection="row"
              marginTop="20px"
              style={{
                border: '1px solid #2B4D6F',
                borderRadius: '10px',
                boxSizing,
              }}
              height="73vh"
              justifyContent="space-between"
            >

              <Box
                style={{
                  padding: "20px",
                  backgroundColor: '#F4F5F6',
                  borderRadius: '10px',
                  boxSizing,
                }}>
                <Link href="/organizations">
                  <Button
                    style={{
                      minWidth: '40px',
                      height: '45px',

                    }}
                    size="medium"
                    variant="contained"
                  >
                    {to('title-create')}
                  </Button>
                </Link>

                <Box
                  paddingTop='2em'
                  paddingBottom='2em'
                  height="100%"
                >
                  <ScrollableList organizationList={organizationList} onSelect={handleSelectedOrganization} />
                </Box>

              </Box>

              <Box
                width="70%"
                margin="20px"
                sx={{ overflow: 'auto' }}
                paddingRight="1em"
                position="relative"
              >

                <Box
                  width="100%"
                  display="flex"
                  alignItems="center"
                  justifyContent="end"
                  marginBottom='20px'
                >
                  <Link href="/products/create">
                    <Button
                      style={{
                        minWidth: '40px',
                        height: '45px',

                      }}
                      size="large"
                      variant="contained"
                    >
                      {tp('title-create')}
                    </Button>
                  </Link>
                </Box>

                <TextField
                  fullWidth
                  label={tp('search')}
                  variant="outlined"
                  onChange={(e) => handleProductFilter(e.target.value)}
                  sx={{ mb: 2 }}
                  data-testid="search-input"
                />

                <Box
                  display="flex"
                  flexDirection="column"
                  marginTop="20px"
                  justifyContent="space-around"
                  width="100%"
                  alignContent="center"
                >
                  {filteredProducts?.map((product) => (
                    <Box
                      paddingBottom="2em"
                    >
                      <CardNavigation
                        key={product.id}
                        id={product.id}
                        name={product.name}
                        description={product.description}
                        url={`/products/${currentOrganization?.id}-${product?.id}-${product?.name}`}
                      />
                    </Box>

                  ))}
                  {
                    filteredProducts?.length === 0 &&
                    <Box
                      display="flex"
                      alignItems="center"
                      justifyContent="center"
                      bgcolor="rgba(223, 142, 22, 0.22)"
                      color="#DF8E16"
                      padding={2}
                      borderRadius={2}
                      width="fit-content"
                      marginTop="15%"
                      alignSelf="center"
                    >
                      <WarningIcon sx={{ color: '#e08c14', marginRight: 1 }} />
                      <Typography color="black">{tp('not-found')}</Typography>
                    </Box>
                  }
                </Box>
              </Box>
            </Box>

          }
        </Box>
      </Container >
    </>
  );
};

Products.getLayout = getLayout;

export default Products;
