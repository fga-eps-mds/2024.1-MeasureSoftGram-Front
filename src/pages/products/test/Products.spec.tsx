import React from 'react';
import { render, screen } from '@testing-library/react';
import { ProductProvider } from '@contexts/ProductProvider';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import Products from '../Products';
import '@testing-library/jest-dom';


jest.mock('next/router', () => ({
  useRouter: () => ({
    push: () => jest.fn(),
  })
}));

describe('Products', () => {
  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <OrganizationProvider>
          <ProductProvider>
            <Products />
          </ProductProvider>
        </OrganizationProvider>
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Renderização e Estilos', () => {
    it('Renderiza corretamente os elementos', () => {
      render(
        <OrganizationProvider>
          <ProductProvider>
            <Products />
          </ProductProvider>
        </OrganizationProvider>
      );

      const organizationBox = screen.getByTestId('organization-box');
      expect(organizationBox).toBeInTheDocument();
      expect(organizationBox).toHaveStyle({
        marginTop: '40px',
        padding: '20px 36px',
        overflowX: 'auto',
        overflowY: 'hidden',
        position: 'relative',
        maxHeight: '120px',
      });
    });
  });
});
