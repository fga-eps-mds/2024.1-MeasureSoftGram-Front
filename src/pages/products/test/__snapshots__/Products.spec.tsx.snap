// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        style="width: 100%;"
      >
        <div
          class="MuiBox-root css-av4yu3"
        >
          <div
            class="MuiBox-root css-g8qhew"
            data-testid="organization-box"
            style="max-height: 120px;"
          >
            <div
              class="MuiBox-root css-o4ooe"
              style="gap: 1rem; padding: 0px; margin-left: -36px; box-sizing: border-box;"
            />
          </div>
          <div
            class="MuiBox-root css-0"
            style="border: 1px solid #113d4c; border-radius: 10px; box-sizing: border-box;"
          >
             
            <div
              class="MuiBox-root css-j7qwjs"
            >
              <div
                class="MuiBox-root css-z5fpi6"
              >
                <div
                  class="MuiBox-root css-0"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                    style="width: 300px; height: 65px;"
                  />
                </div>
              </div>
              <div
                class="MuiBox-root css-1182l2u"
              >
                <div
                  class="MuiBox-root css-lvyu5j"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                    style="width: 250px; height: 60px;"
                  />
                </div>
                <span
                  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-wkdwxb-MuiSkeleton-root"
                  style="width: 40px; height: 40px;"
                />
              </div>
            </div>
             
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      style="width: 100%;"
    >
      <div
        class="MuiBox-root css-av4yu3"
      >
        <div
          class="MuiBox-root css-g8qhew"
          data-testid="organization-box"
          style="max-height: 120px;"
        >
          <div
            class="MuiBox-root css-o4ooe"
            style="gap: 1rem; padding: 0px; margin-left: -36px; box-sizing: border-box;"
          />
        </div>
        <div
          class="MuiBox-root css-0"
          style="border: 1px solid #113d4c; border-radius: 10px; box-sizing: border-box;"
        >
           
          <div
            class="MuiBox-root css-j7qwjs"
          >
            <div
              class="MuiBox-root css-z5fpi6"
            >
              <div
                class="MuiBox-root css-0"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                  style="width: 300px; height: 65px;"
                />
              </div>
            </div>
            <div
              class="MuiBox-root css-1182l2u"
            >
              <div
                class="MuiBox-root css-lvyu5j"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                  style="width: 250px; height: 60px;"
                />
              </div>
              <span
                class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-wkdwxb-MuiSkeleton-root"
                style="width: 40px; height: 40px;"
              />
            </div>
          </div>
           
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
